### PR TITLE
First implementation

### PR DESCRIPTION
**Pull request by darosior**
This implements the `revault` lib, a generalistic (script-config wise) library for creating Revault transactions throughout our binaries.

## Aim
- Being flexible on what *may* change in the future, i.e. Script policies. [Miniscript](http://bitcoin.sipa.be/miniscript/) allows us to do so by failing early (at policy-to-miniscript compilation) on malleable or unsafe scripts, before we even create an output we could send funds to. Obviously, doing so without miniscript is not an option.
- Being restrictive on what *must* not change in the (foreseeable) future, i.e. transaction structure. I used a newtype pattern around rust-bitcoin's Transaction for each Revault transaction type ("Vault", Unvault, Spend, Cancel, Emergency). This results in some boilerplate code in the library but i think this can helps us avoiding mistakes when using it and going forward with transaction puzzles.

## Structure
See the doc, but there are 3 parts:
- The errors
- The transactions (naked, no script just the structure)
- The scripts (miniscript logic)

The miniscript logic spills out a bit in the transactions mod with the satisfier, but it made sense to me to have the txin satisfier part of the tx module.

## Hacks
- We need BIP143 sighash and rust-bitcoin 0.24 isn't released yet so i used temporary branches for rust-miniscript and rust-bitcoin (the former would get angry if we took the latter directly out of the Github master branch).

## Still TODO
- More test coverage:
  - Some erroring path are still untested
  - Fuzz testing soon (:tm:)
- Extended public keys support (needs something like https://github.com/rust-bitcoin/rust-miniscript/pull/116 or https://github.com/rust-bitcoin/rust-miniscript/pull/93)
- Maybe PSBT support

Finally, note that it was not intended to be a PR at first so the commit history is "sketchy" so you are probably better reviewing the whole diff instead ...

### CODE REVIEW

File: src/transations.rs [hunk 1]
```diff
@@ -0,0 +1,1359 @@
...
+            None
+        }
+
+        match *self {
+            RevaultTransaction::VaultTransaction(ref tx)
+            | RevaultTransaction::UnvaultTransaction(ref tx)
+            | RevaultTransaction::SpendTransaction(ref tx)
+            | RevaultTransaction::CancelTransaction(ref tx)
+            | RevaultTransaction::EmergencyTransaction(ref tx)
+            | RevaultTransaction::FeeBumpTransaction(ref tx) => {
```
Comment (edouardparis) [inline src/transations.rs:430]: what do you think of implementing a method for VaultTransaction to get the tx ? I think the pattern matching will be often repetitive

Comment (darosior) [reply src/transations.rs:430]: Yeah, it is (but at least only on the inside). I was thinking of a macro but a method is OK too

File: src/transations.rs [hunk 2]
```diff
@@ -0,0 +1,1359 @@
...
+        (
+            (managers_priv, managers),
+            (non_managers_priv, non_managers),
+            (cosigners_priv, cosigners),
+        )
+    }
+
+    // Routine for ""signing"" a transaction
+    fn satisfy_transaction_input(
+        secp: &secp256k1::Secp256k1<secp256k1::All>,
```
Comment (edouardparis) [inline src/transations.rs:701]: `secp256k1::PublicKey::from_secret_key` and `secp.sign` only needs `secp256::Signing` Trait, I think you can change for:
```rust
 fn satisfy_transaction_input<C: Signing>(
        secp: &secp256k1::Secp256k1<C>,
```

Comment (edouardparis) [reply src/transations.rs:701]: ah its in the mod tests, I do not really care

Comment (darosior) [reply src/transations.rs:701]: Yeah, it was just a helpful quick routine for the tests: we can generalize it later if we ever need.

File: src/transations.rs [hunk 3]
```diff
@@ -0,0 +1,1359 @@
...
+    /// Create a new cancel transaction.
+    /// A cancel transaction always pays to a vault output and spend the unvault output, and
+    /// may have a fee-bumping input.
+    ///
+    /// # Errors
+    /// - If prevouts or txouts type mismatch.
+    pub fn new_cancel(
+        prevouts: &[RevaultPrevout],
+        txouts: &[RevaultTxOut],
+    ) -> Result<RevaultTransaction, RevaultError> {
```
Comment (JSwambo) [inline src/transations.rs:178]: nit: `new_unvault()` and `new_spend()` return `Result<Self, RevaultError>`, 
while `new_cancel()` and `new_emergency()` return `Result<RevaultTransaction, RevaultError>`. I don't think there's a functional difference but maybe its better to be consistent.

**Review by JSwambo — APPROVED** [review]
Ack

File: src/transations.rs (hunk 3 above)
Comment (darosior) [reply src/transations.rs:178]: Absolutely, thanks!

**Comment by darosior** [comment]
Removed `Self`-isms, and added a commit adding a ref-to-the-inner-tx method

**Comment by darosior** [comment]
Added a commit with more tests..

**Comment by darosior** [comment]
Rebased on master now that i added the CI.

**Review by JSwambo — APPROVED** [review]
Ack!

**Pull request merged**